File: src/classes/PolyTrajPlanner.m
classdef PolyTrajPlanner < handle
    %   third and fifth order polynomial for trajectory plan(for now)
    %   TO DO: extends other style cubic splines
    %   ref: trajectory planning for automatic machines and robots, chapter 4.4  
    %   TO DO: smoothing cubic splines, chapter 4.4.5
    properties
        num
        tf
        dt
        
        poly_params
        order
    end

    
    
    methods
        %% Constructor and Get the Polynomial Parameters
        function obj = PolyTrajPlanner(pos, tf, order)
            obj.order = order;
            n = length(pos);
            obj.num = n;
            if length(tf)==n || length(tf)==1
                obj.tf = tf;
                obj.dt = tf(end)/(n-1);
            else
                error('dimension of positoin and time mismatch');
            end
            obj.poly_params = zeros(order+1, n-1);
            obj.poly_params = obj.PolyParams(pos, tf);
        end
        
        %% Generate the Trajectory
        function [pos, vel, acc] = GenerateTraj(obj, dt)
            pos = []; vel = []; acc = [];
            for t=0:dt:obj.tf(end)
                [p, v, a] = obj.GenerateMotion(t);
                pos = [pos, p];
                vel = [vel, v];
                acc = [acc, a];
            end
        end
        
        function [p, v, a] = GenerateMotion(obj, t)
            if length(obj.tf)==1
                time = 0:obj.dt:obj.tf;
            else
                time = obj.tf;
            end
            idx = discretize(t, time);
            p = obj.PolyPos(t)*obj.poly_params(:,idx);
            v = obj.PolyVel(t)*obj.poly_params(:,idx);
            a = obj.PolyAcc(t)*obj.poly_params(:,idx);
        end
        
        %% Elementary Row Vector for Position, Velocity and Acceleration
        function res = PolyPos(obj, t)
            if obj.order==3
                res = [1, t, t^2, t^3];
            elseif obj.order==5
                res = [1, t, t^2, t^3, t^4, t^5];
            end
        end
        
        function res = PolyVel(obj, t)
            if obj.order==3
                res = [0, 1, 2*t, 3*t^2];
            elseif obj.order==5
                res = [0, 1, 2*t, 3*t^2, 4*t^3, 5*t^4];
            end
        end
        
        function res = PolyAcc(obj, t)
            if obj.order==3
                res = [0, 0, 2, 6*t];
            elseif obj.order==5
                res = [0, 0, 2, 6*t, 12*t^2, 20*t^3];
            end
        end
        
        %% Calculate Polynomial Parameters
        function params = PolyParams(obj, pos, tf)
            if obj.order==3
                if obj.num>2
                    params = obj.PolyContiAcc(pos, tf);
                else
                    params = obj.PolyAutoVel(pos, tf);
                end
            elseif obj.order==5
                rhs = [pos(1); pos(2); 0; 0; 0; 0];
                lhs = obj.LhsMat(0, tf);
                params = lhs\rhs;                
            end
                end
        
        function res = LhsMat(obj, t0, tf)
            if obj.order==3
                res = [obj.PolyPos(t0); obj.PolyPos(tf); obj.PolyVel(t0); obj.PolyVel(tf)];
            elseif obj.order==5
                res = [obj.PolyPos(t0); obj.PolyPos(tf);...
                            obj.PolyVel(t0); obj.PolyVel(tf);...
                            obj.PolyAcc(t0); obj.PolyAcc(tf)];
            end
        end
    
        function params = PolyAutoVel(obj, pos ,tf)
            % calculate the velocity heuristically 
            t = 0:obj.dt:tf;
            n = obj.num;
            for idx=1:n-1
                vel(1) = 0;
                if idx==n-1
                    vel(idx+1) = 0;
                else
                    k1 = (pos(idx+1)-pos(idx))/obj.dt;
                    k2 = (pos(idx+2)-pos(idx+1))/obj.dt;
                    if sign(k1*k2)==1
                        vel(idx+1) = 0.5*(k1+k2);
                    else
                        vel(idx+1) = 0;
                    end
                end
                rhs = [pos(idx); pos(idx+1); vel(idx); vel(idx+1)];
                lhs = obj.LhsMat(t(idx), t(idx+1));
                params(:,idx) = lhs\rhs;
            end                         
        end
        
        function params = PolyContiAcc(obj, pos, tf)
            % make the acceleration continuous
            n = obj.num;
            rhs = zeros(4*(n-1), 1);
            lhs = zeros(4*(n-1), 4*(n-1));
            if length(obj.tf)==1                
                t = 0:obj.dt:tf;
            else
                t = tf;
            end
            if obj.num>2
                rhs(1) = pos(1);
                rhs(2*(n-1)) = pos(n);
                lhs(2*(n-1)+1, 1:4) = obj.PolyVel(t(1));
                lhs(2*(n-1)+n, 4*n-7:4*n-4) = obj.PolyVel(t(n));
                rhs(2*(n-1)+1) = 2;
                rhs(2*(n-1)+n) = -3;
                for idx=2:n-1
                    rhs(2*idx-2) = pos(idx);
                    rhs(2*idx-1) = pos(idx);
                    lhs(2*(n-1)+idx, 4*idx-7:4*idx-4) = obj.PolyVel(t(idx));
                    lhs(2*(n-1)+idx, 4*idx-3:4*idx) = -obj.PolyVel(t(idx));
                    lhs(3*n-3+idx, 4*idx-7:4*idx-4) = obj.PolyAcc(t(idx));
                    lhs(3*n-3+idx, 4*idx-3:4*idx) = -obj.PolyAcc(t(idx));
                end
                
                for idx=1:n-1
                    lhs(2*idx-1, 4*idx-3:4*idx) = obj.PolyPos(t(idx));
                    lhs(2*idx, 4*idx-3:4*idx) = obj.PolyPos(t(idx+1));
                end
                params = lhs\rhs;
                params = reshape(params, 4, n-1);
            else
                error('error input position');
            end
            
        end
        
        %% Plot Function for Test and Presentation
        function PlotAVP(obj, dt)
            [pos, vel, acc] = obj.GenerateTraj(dt);
            t = 0:dt:obj.tf(end);
            motion_data = [pos; vel; acc];
            motion_name = {'position', 'velocity', 'acceleration'};
            for idx=1:3
                subplot(3, 1, idx);
                plot(t, motion_data(idx, :)); grid on
                ylabel(motion_name{idx});
            end                       
        end
        
    end
    
    
    
end

File: src/cfiles/PolyTrajPlanner.cpp
#include "stdafx.h"
#include "PolyTrajPlanner.h"


PolyTrajPlanner::PolyTrajPlanner(VectorXd pos, double tf, int order)
{
	_order = order;
	int n = pos.size();
	_nump = n;
	_tf = tf;
	_tf_vec.setZero(n);
	_dt = tf / (double)(n - 1);
	_poly_params.setZero(order + 1, n - 1);
	_poly_params = CalcPolyParams(pos);
}

PolyTrajPlanner::PolyTrajPlanner(VectorXd pos, double* tf_vec, int order)
{
	_order = order;
	int n = pos.size();
	_nump = n;
	_tf_vec.setZero(n);
	for (int idx = 0; idx < n; idx++)
	{
		_tf_vec(idx) = tf_vec[idx];
	}
	_tf = 0;
	_dt = 0;
	_poly_params.setZero(order + 1, n - 1);
	_poly_params = CalcPolyParams(pos);
}

RowVectorXd PolyTrajPlanner::PolyPos(double t)
{
	RowVectorXd res;
	if (_order == 3)
	{		
		res.setZero(4);
		res << 1.0, t, pow(t,2), pow(t,3);
	}
	else if (_order == 5)
	{
		res.setZero(6);
		res << 1, t, pow(t,2), pow(t,3), pow(t,4), pow(t,5);
	}
		
	return res;
}

RowVectorXd PolyTrajPlanner::PolyVel(double t)
{
	RowVectorXd res;
	if (_order == 3)
	{
		res.setZero(4);
		res << 0, 1, 2 * t, 3 * pow(t, 2);
	}
	else if (_order == 5)
	{
		res.setZero(6);
		res << 0, 1, 2 * t, 3 * pow(t, 2), 4 * pow(t, 3), 5 * pow(t, 4);
	}
	return res;
}

RowVectorXd PolyTrajPlanner::PolyAcc(double t)
{
	RowVectorXd res;
	if (_order == 3)
	{
		res.setZero(4);
		res << 0, 0, 2, 6 * t;
	}
	else
	{
		res.setZero(6);
		res << 0, 0, 2, 6 * t, 12 * pow(t, 2), 20 * pow(t, 3);
	}
	return res;
}

MatrixXd PolyTrajPlanner::LhsMat(double t0, double tf)
{
	MatrixXd res;
	if (_order == 3)
	{
		res.setZero(4, 4);
		res.row(0) = PolyPos(t0);
		res.row(1) = PolyPos(tf);
		res.row(2) = PolyVel(t0);
		res.row(3) = PolyVel(tf);
	}
	else if(_order == 5)
	{
		res.setZero(6, 6);
		res.row(0) = PolyPos(t0);
		res.row(1) = PolyPos(tf);
		res.row(2) = PolyVel(t0);
		res.row(3) = PolyVel(tf);
		res.row(4) = PolyAcc(t0);
		res.row(5) = PolyAcc(tf);
	}
	return res;
}

MatrixXd PolyTrajPlanner::CalcPolyParams(VectorXd pos)
{
	MatrixXd params;
	if (_order == 3)
	{
		if (_nump > 2)
		{
			params = PolyContiAcc(pos);
		}
		else
		{
			params = PolyAutoVel(pos);
		}
	}
	else if (_order==5)
	{
		params.setZero(6,1);
		Matrix<double, 6, 1> rhs;
		rhs << pos(0), pos(1), 0, 0, 0, 0;
		MatrixXd lhs = LhsMat(0, _tf);
		params = lhs.lu().solve(rhs);
	}
	return params;
}

MatrixXd PolyTrajPlanner::PolyAutoVel(VectorXd pos)
{
	MatrixXd params;
	params.setZero(4, _nump - 1);
	VectorXd vel, t;
	vel.setZero(_nump);
	t.setLinSpaced(_nump, 0, _tf);
	for (int idx = 0; idx < _nump - 1; idx++)
	{
		vel(0) = 0;
		if (idx == _nump - 2)
		{
			vel(idx + 1) = 0;
		}
		else
		{
			double k1 = (pos(idx + 1) - pos(idx)) / _dt;
			double k2 = (pos(idx + 2) - pos(idx + 1)) / _dt;
			if (MathTools::Sign(k1*k2) == 1)
			{
				vel(idx + 1) = 0.5*(k1 + k2);
			}
			else
			{
				vel(idx + 1) = 0;
			}
		}
		Vector4d rhs;
		rhs << pos(idx), pos(idx + 1), vel(idx), vel(idx + 1);
		MatrixXd lhs = LhsMat(t(idx), t(idx + 1));
		params.col(idx) = lhs.lu().solve(rhs);
	}
	return params;
}

MatrixXd PolyTrajPlanner::PolyContiAcc(VectorXd pos)
{
	int n = _nump;
	MatrixXd params, rhs, lhs;
	params.setZero(4 * (n - 1), 1);
	rhs.setZero(4 * (n - 1), 1);
	lhs.setZero(4 * (n - 1), 4 * (n - 1));
	VectorXd t;
	if (fabs(_tf) < EPS)
	{
		t = _tf_vec;
	}
	else
	{
		t.setLinSpaced(n, 0, _tf);
	}

	rhs(0) = pos(0);
	rhs(2 * (n - 1)-1) = pos(n-1);
	lhs.block(2 * (n - 1), 0, 1, 4) = PolyVel(t(0));
	lhs.block(2 * (n - 1) + n - 1, 4 * n - 8, 1, 4) = PolyVel(t(n - 1));
	for (int idx = 1; idx < n - 1; idx++)
	{
		rhs(2 * idx - 1) = pos(idx);
		rhs(2 * idx) = pos(idx);
		lhs.block(2 * (n - 1) + idx, 4 * idx - 4, 1, 4) = PolyVel(t(idx));
		lhs.block(2 * (n - 1) + idx, 4 * idx, 1, 4) = -PolyVel(t(idx));
		lhs.block(3 * n - 3 + idx, 4 * idx - 4, 1, 4) = PolyAcc(t(idx));
		lhs.block(3 * n - 3 + idx, 4 * idx, 1, 4) = -PolyAcc(t(idx));

	}
	for (int idx = 0; idx < n - 1; idx++)
	{
		lhs.block(2 * idx, 4 * idx, 1, 4) = PolyPos(t(idx));
		lhs.block(2 * idx + 1, 4 * idx, 1, 4) = PolyPos(t(idx + 1));
	}
	params = lhs.lu().solve(rhs);
	params.resize(4, n - 1);

	return params;
}


double PolyTrajPlanner::GenerateMotion(double t)
{
	VectorXd time_vec;
	if (fabs(_tf) < EPS)
	{
		time_vec = _tf_vec;
	}
	else
	{
		time_vec.setLinSpaced(_nump, 0, _tf);
	}	
	int idx = MathTools::Discretize(time_vec, _nump, t);
	double p = PolyPos(t)*_poly_params.col(idx);

	return p;
}

PolyTrajPlanner::~PolyTrajPlanner()
{
}
